experimental/dds/tree2/src/feature-libraries/modular-schema/typedSchema/buildViewSchemaCollection.ts

```diff
@@ -0,0 +1,216 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { GlobalFieldKey, TreeSchemaIdentifier } from "../../../core";
+import { SchemaCollection } from "../view";
+import { fail } from "../../../util";
+import { defaultSchemaPolicy } from "../../defaultSchema";
+import { FieldKinds, forbidden, value } from "../../defaultFieldKinds";
+import { SchemaLibraryData, SourcedAdapters } from "./schemaBuilder";
+import { FieldSchema, GlobalFieldSchema, TreeSchema, allowedTypesIsAny } from "./typedTreeSchema";
+import { normalizeFlexListEager } from "./flexList";
+
+// TODO: tests for this file
+
+/**
+ * Build and validate a SchemaCollection.
+ *
+ * As much as possible tries to detect anything that might be a mistake made by the schema author.
+ * This will error on some valid but probably never intended to be used patterns
+ * (like libraries with the same name, nodes which are impossible to create etc).
+ */
+export function buildViewSchemaCollection(
+	libraries: readonly SchemaLibraryData[],
+): SchemaCollection {
+	const globalFieldSchema: Map<GlobalFieldKey, GlobalFieldSchema> = new Map();
+	const treeSchema: Map<TreeSchemaIdentifier, TreeSchema> = new Map();
+	const adapters: SourcedAdapters = { tree: [], fieldAdapters: new Map() };
+
+	const errors: string[] = [];
+	const librarySet: Set<SchemaLibraryData> = new Set();
+	const libraryNames: Set<string> = new Set();
+
+	for (const library of libraries) {
+		if (librarySet.has(library)) {
+			// SchemaBuilder should ensure this doesn't happen, but include it here for completeness.
+			errors.push(`Duplicate library named "${library.name}"`);
+			continue;
+		}
+		librarySet.add(library);
+		if (libraryNames.has(library.name)) {
+			// This wouldn't break anything, but could make error messages confusing, so its better to avoid duplicates.
+			errors.push(`Found another library with name "${library.name}"`);
+		}
+
+		for (const [key, field] of library.globalFieldSchema) {
+			assert(field.builder.name === library.name, "unexpected builder on field");
```

Comment:
the global field remembers which builder created it, and if its part of a different library thats bad. It would make error messages about it wrong, and probably indicates something got messed up somehow. The risk of incorrect error messages is the main reason I checked this case.